eventuals/notification.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include "eventuals/lock.h"
+#include "eventuals/then.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+// This header file defines a 'Notification' abstraction, which allows
+// an eventual to be able to wait for a single occurrence of a single
+// event. Said another way, the 'Notification' object maintains a
+// private boolean "notified" that transitions to 'true' at most once.
+//
+// This is similar to 'absl::Notification'
+// (https://abseil.io/docs/cpp/guides/synchronization#notification).
+//
+// You can call 'Wait()' on a 'Notification' to wait until the
+// "notified" state is 'true' and 'Notify()' to set the notification's
+// "notified" state to 'true' and notify all waiting eventuals that
+// the event has occurred. This method should only be called once and
+// will do nothing on repeated calls.
+//
+// Note that while 'Notify()' should only be called once, it is
+// perfectly valid to call 'Wait()' multiple times and/or from
+// multiple eventuals -- even after the notification's "notified"
+// state has been set -- in which case those calls will not wait.
+//
+// Note that the lifetime of a 'Notification' requires careful
+// consideration; it might not be safe to destroy a 'Notification'
+// after calling 'Notify()' since other eventuals may have called, or
+// will call, 'Wait()'.
+class Notification final : public Synchronizable {
+ public:
+  Notification()
+    : notification_(&lock()) {}
+
+  ~Notification() override = default;
+
+  [[nodiscard]] auto Notify() {
+    return Synchronized(Then([this]() {
+      if (!notified_) {
+        notified_ = true;
+        notification_.NotifyAll();
+      }
+    }));
+  }
+
+  [[nodiscard]] auto Wait() {
```

Comment:
Please add comments for non-trivial public API methods.

Here's my take, inspired by `absl`'s: https://github.com/abseil/abseil-cpp/blob/149ab40581d33febf3f8bddb52505c9ee9edd838/absl/synchronization/notification.h#L87

```
// Continues once the notification's "notification" state is true, which might be immediately
// if Notify() has already been called.
```